Re-focus on parent surface if it is available

BrassyPanache opened the pull request:
First I must say that I am not sure if my approach is sound.

My primary issue was IntelliJ IDEA's code suggestion pop-up not returning focus
to the active editing window.

I have spent some time looking at the changes of @Xyene (#5398) and
@RyanDwyer (#2103). I think my proposed change maintains the status
quo for the most part whilst fixing my focus issue.

I have verified that @Xyene's fix for IntelliJ sub-menus still works.

I have done basic testing which consists of:

- Chrome
- IntelliJ IDEA 2020.2.1
- VSCode
- Alacritty

It seems to hold up. I at least didn't see any obvious errors.

Relates to #3007

Xyene commented on sway/desktop/xwayland.c:
```diff
@@ -92,22 +92,12 @@ static void unmanaged_handle_unmap(struct wl_listener *listener, void *data) {
-				return;
-			}
+	if (seat->wlr_seat->keyboard_state.focused_surface == xsurface->surface) {
+		// This simply returns focus to the parent surface if there's one avaialble.
```
Small nit here.

```suggestion
		// This simply returns focus to the parent surface if there's one available.
```

Xyene commented on sway/desktop/xwayland.c:
```diff
@@ -92,22 +92,12 @@ static void unmanaged_handle_unmap(struct wl_listener *listener, void *data) {
+	if (seat->wlr_seat->keyboard_state.focused_surface == xsurface->surface) {
+		// This simply returns focus to the parent surface if there's one avaialble.
+		// This seems to handle JetBrains issues.
+		if (xsurface->parent && xsurface->parent->surface) {
```
I think this should still be calling `wlr_xwayland_or_surface_wants_focus`, does doing so break this?

BrassyPanache replied:
I've just tested a new build with this suggestion and all seems to work as expected. I'm not sure that I've done it all correctly as I'm new to the code base :smile:

BrassyPanache commented:
Thanks for the super quick review.  In my testing I added quite a bit of logging to verify that the same surfaces were given focus.  As far as I could tell it was the case.  I removed the additional logging before I committed.  I've attached one of my annotated [logs](https://github.com/swaywm/sway/files/5134103/sway.1.log) for reference.

Xyene commented on sway/desktop/xwayland.c:
```diff
@@ -92,20 +92,15 @@ static void unmanaged_handle_unmap(struct wl_listener *listener, void *data) {
+		// This simply returns focus to the parent surface if there's one available.
+		// This seems to handle JetBrains issues.
+		if (xsurface->parent &&
+			wlr_surface_is_xwayland_surface(xsurface->parent->surface)) {
```
```suggestion
				wlr_surface_is_xwayland_surface(xsurface->parent->surface)) {
```

(Otherwise it's easy to miss that this statement is part of the if condition and not its body.)

BrassyPanache replied:
Thanks again for the review. This should now be resolved. I've also rebased and squished into a single commit.

Xyene reviewed (commented):
Thanks, apart from a small nit this looks good to me.

Also, don't forget to rebase + squash your commits into one, then we can merge this.

Xyene reviewed (approved):
Thanks!
